File: content/problems/0841-keys-and-rooms/solutions/graph-reachability-dfs/Solution.java
class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms){
        boolean[] seen=new boolean[rooms.size()];
        Deque<Integer> stack=new ArrayDeque<>();
        seen[0]=true;
        stack.push(0);
        int visited=0;
        while(!stack.isEmpty()){
            int room=stack.pop();
            visited++;
            for(int key:rooms.get(room))if(!seen[key]){
                seen[key]=true;
                stack.push(key);
            }
        }
        return visited==rooms.size();
    }
}

File: content/problems/0841-keys-and-rooms/solutions/graph-reachability-dfs/Solution.kt
class Solution {
    fun canVisitAllRooms(rooms:List<List<Int>>):Boolean{
        val seen=BooleanArray(rooms.size);
        val stack=java.util.ArrayDeque<Int>();
        seen[0]=true;
        stack.push(0);
        var visited=0;
        while(stack.isNotEmpty()){
            val room=stack.pop();
            visited++;
            for(key in rooms[room])if(!seen[key]){
                seen[key]=true;
                stack.push(key)
            }
        };
        return visited==rooms.size
    }
}
